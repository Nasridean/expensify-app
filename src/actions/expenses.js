import { v4 as uuidv4 } from 'uuid';
export const addExpense = ({ description = '', createdAt = 0, amount = 0, note = '' } = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuidv4(),
        description,
        createdAt,
        amount,
        note
    }
});
export const removeExpense = ({ id }) => ({
    type: 'REMOVE_EXPENSE',
    id
});
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
}); 