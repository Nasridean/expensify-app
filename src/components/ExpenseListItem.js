import React from 'react';
import { NavLink } from 'react-router-dom';

export const ExpenseListItem =  (props) => (
    <div>
        <NavLink to={`/edit/${props.expense.id}`} activeClassName="is-active">{props.expense.description}</NavLink>
        <p>{props.expense.amount}</p>
    </div>
);


export default ExpenseListItem;