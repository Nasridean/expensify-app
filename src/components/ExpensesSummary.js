import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = (props) => (
    <div className="page-header">
        <div className="content-container">
            <h1 className="page-header__title">Total of <span>{props.expenses.length}</span> {props.expenses.length > 1 ? 'expenses:' : 'expense:'} <span>{numeral(selectExpensesTotal(props.expenses)/100).format('$0,0.00')}</span></h1>
            <div className="page-header__actions">
                <Link className="button" to="/create">Add Expense</Link>
            </div>
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filter),
    };
};

export default connect(mapStateToProps)(ExpensesSummary);