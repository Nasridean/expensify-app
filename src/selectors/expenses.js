import moment from "moment";

export default (expenses, { text, sortBy, startDate, endDate }) => {
    console.log(expenses);
    return expenses.filter(( expense )=> {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = typeof text !== 'string' || text.length === 0 || expense.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return  b.createdAt - a.createdAt;
        } else if (sortBy === 'amount') {
            return b.amount - a.amount;
        } else {
            return 0;
        }
    });  
};