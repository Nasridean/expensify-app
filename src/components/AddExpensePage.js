import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense, addExpense } from '../actions/expenses';
export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
        AddExpensePage
        <ExpenseForm
            onSubmit={this.onSubmit}
        />
    </div>
        );
    };
};
const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});
export default connect(undefined, mapDispatchToProps)(AddExpensePage);