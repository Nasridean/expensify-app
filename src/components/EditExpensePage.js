import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
export class EditExpensePage extends React.Component {
    onSubmit = ((expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    });
    onClick = () => {
        this.props.startRemoveExpense(this.props.expense);
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
        <ExpenseForm 
            expense ={this.props.expense}
            onSubmit={this.onSubmit}
        />
        <button onClick={this.onClick}>Remove</button>
    </div>
        )
    }
}
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => props.match.params.id === expense.id)
    }
}
const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (expense) => dispatch(startRemoveExpense(expense))
})
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);