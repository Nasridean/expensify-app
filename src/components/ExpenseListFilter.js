import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilter extends React.Component {
    state = {
        calendarFocused: null
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };
    onSortByChange = (e) => {
            e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();
    };
    onTextChange = (e) => {
            this.props.setTextFilter(e.target.value);
    };
    render() {
        return (
            <div>
                <input type="text" value={this.props.filter.text} onChange={this.onTextChange} />
                <select value={this.props.filter.sortBy} onChange={this.onSortByChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filter.startDate}
                    endDate={this.props.filter.startDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        );
    };
}; 

const mapStateToProps = (state) => {
    return {
        filter: state.filter
    };
};
const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setEndDate: (startDate) => dispatch(setEndDate(startDate)),
    setStartDate: (endDate) => dispatch(setStartDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);