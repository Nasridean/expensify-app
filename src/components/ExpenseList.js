import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {props.expenses.map((expense) => <ExpenseListItem expense={expense} key={expense.id} />)}
        <h4>Total of {props.expenses.length} {props.expenses.length > 1 ? 'expenses:' : 'expense:'} </h4>{numeral(selectExpensesTotal(props.expenses)/100).format('$0,0.00')}
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filter),
    };
};

export default connect(mapStateToProps)(ExpenseList);