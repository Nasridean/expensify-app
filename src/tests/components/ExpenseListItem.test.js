import React from 'react';
import expenses from '../fixtures/expenses';
import { shallow } from 'enzyme';
import { ExpenseListItem } from '../../components/ExpenseListItem';

test('should render an expense item with some expense data', () => {
    const wrapper = shallow(<ExpenseListItem expense={expenses[1]} />);
    expect(wrapper).toMatchSnapshot();
})