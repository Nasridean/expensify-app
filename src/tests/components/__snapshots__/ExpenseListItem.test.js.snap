// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an expense item with some expense data 1`] = `
<div>
  <NavLink
    activeClassName="is-active"
    to="/edit/1"
  >
    House
  </NavLink>
  <p>
    100
  </p>
</div>
`;
